### PR TITLE
stage1: Implement @reduce builtin for vector types

### LINKED ISSUE #2698: @reduce builtin
[Accepted Proposal](https://github.com/ziglang/zig/issues/2698#issuecomment-533727296)

----

https://github.com/ziglang/zig/issues/903#issuecomment-492300314

Branching on SIMD is complicated.[1] If only some of the values match you have to both execute the `if` and the `else` block, and merge the results. I think this should be explicit, which `@Vector(_, bool)` (which all vector comparisons return, although I do not fully understand the difference in zig between `u1` and `bool`) having built-in functions that return `bool`: `.all()` and `.any()`.

You might then run some scalar code, and if that code is `pure` then the compiler can still safely do optimizations that result in it being processed possibly twice. So in addition to a `comptime` block I think we also need a `pure` block that would give a compile error if the block is not `pure`.[2]

[1] Here is HSAIL talking about wavefront scheduling http://www.hsafoundation.com/html_spec111/HSA_Library.htm#PRM/Topics/02_ProgModel/wavefronts_lanes_and_wavefront_sizes.htm%3FTocPath%3DHSA%25C2%25A0Programmer's%2520Reference%2520Manual%2520Version%25201.1.1%2520%7CChapter%25202.%2520HSAIL%2520Programming%2520Model%7C2.6%2520Wavefronts%252C%2520Lanes%252C%2520and%2520Wavefront%2520Sizes%7C_____0
[2] I don't want to add too many keywords. Perhaps we could have attribute that can be attached to blocks, functions, variables, and other things that are just `enum` literals? (This is also very similar to C where _ is reserved.)

```
.comptime {
   foo();
}
.pure {
   bar();
}
fn extern .luacc call_my_lua_func(f: u32);
```
```

### PR DESCRIPTION
**Pull request by LemonBoy**
The builtin folds a Vector(N,T) into a scalar T using a specified operator.

Closes #2698

### CODE REVIEW

**Comment by andrewrk** [comment]
My goodness, we just cannot catch a break with that windows CI

File: lib/std/builtin.zig [hunk 1]
```diff
@@ -98,6 +98,16 @@ pub const AtomicOrder = enum {
...
 };
 
+/// This data structure is used by the Zig language code generation and
+/// therefore must be kept in sync with the compiler implementation.
+pub const ReduceOp = enum {
+    And,
+    Or,
+    Xor,
+    Min,
+    Max,
```
Comment (andrewrk) [inline lib/std/builtin.zig:108]: [lower case tag names](https://ziglang.org/documentation/master/#Style-Guide) for new enums please. I normally don't want to care about style stuff but I also don't want to have to make unnecessary breaking changes later.

**Review by andrewrk — CHANGES_REQUESTED** [review]
Nice progress! 1 nitpick and then this is good to land.

BTW why not work towards this in stage2? This is still useful progress because the changes to builtin.zig, behavior tests, and zig_llvm.{h,cpp} will persist, but I have a feeling you would enjoy the work more on stage2, and once we get far enough on that, the current plan is to delete the stage1 codebase (#6378).

Also I'm happy to do the nitpick myself in the merge process if you want, just say the word.

**Comment by LemonBoy** [comment]
> BTW why not work towards this in stage2? 

Because I was playing with SIMD rather than with the compiler :^)
When the LLVM backend for stage2 is kickstarted I'll definitely move to the Zig side of the moon.

> Also I'm happy to do the nitpick myself in the merge process if you want, just say the word.

Yeah go ahead, didn't know about the lowercase rule for enum members.

**Comment by andrewrk** [comment]
Oh, hmm. Gonna change my mind on this one. Capitalized enum tag names in this case works around the conflicts with the language keywords. :shrug:

**Comment by andrewrk** [comment]
>  When the LLVM backend for stage2 is kickstarted I'll definitely move to the Zig side of the moon.

Alright well that's certainly motivating for me or someone else to get started on  #6541 :-)

**Comment by LemonBoy** [comment]
> Can you implement .Add and .Mul, please?

Sure, I didn't implement those because I wasn't sure about the names to use (nor I had any need for those ops).
Since the add/sub/mul provided by LLVM are wrapping I'd go for `Wrapping{Add,Mul}`.

**Comment by data-man** [comment]
@LemonBoy 
> Sure

Before 0.7.0 release? :)

But I thought about integer/float overflow for `.Add` and `.Mul`. :(
E.g. if a vector is `float32` type, can `@reduce` return `float64`?

**Pull request merged**